Advanced styles customization

earshinov opened the pull request:
Hello @madoar ,

It is usually considered a good practice of visual design to restrict colors used in the application UI to a limited subset which might be called a palette or a theme. In some cases, such theme lists not only base colors, but also their shades from light to dark to be used in different parts of the UI.

Version 5.0.0 of angular-archwizard exposes SCSS variables which can be used for customizing base colors of the wizard. However, things like using `darken(…, 10%)` to derive a hover color from a base color are still hard coded and hard for a library user to override.

This PR extends angular-archwizard's style customization support. Instead of a small number of variables controlling base colors (`$wz-color-default`, …, `$wz-color-editing`) it exposes a complex `$wz-colors` structure that completely defines a color scheme to be used for navbar indicators, including colors for hovered state.

As an auxiliary change, the updated `*.scss` file also exposes `$wz-layouts` and `$wz-indicator-styles` variables. Library users can override these to exclude unused layouts and indicator style and thus optimize CSS output size.

Notes:
- I have been using this modified version of angular-wizard in order to customize its colors for our project.
- This PR introduces incompatible changes for those who already customize wizard's appearance using SCSS. For other users, changes will go unnoticed. As a matter of fact, this PR does not introduce _any_ meaningful changes in the CSS file included into angular-archwizard npm package.

<a href="https://gitpod.io/#https://github.com/madoar/angular-archwizard/pull/235"><img src="https://gitpod.io/api/apps/github/pbs/github.com/earshinov/angular-archwizard.git/65ba9406a3c02f8bd78029f78fef0a2290f02253.svg" /></a>

madoar commented on src/css/archwizard.scss:
```diff
@@ -1,9 +1,97 @@
+// Override this variable to exclude indicator styles you don't use.
+$wz-indicator-styles: 'small', 'large-filled', 'large-empty', 'large-filled-symbols', 'large-empty-symbols' !default;
+
+// Color definitions
```
Does it maybe make sense to use a different comment style here? I think something like:

```scss
/**
 * ...
 * ...
 */
```

may be easier to read, but I don't know any conventions for comments in scss

earshinov replied:
Hmm... It looks like SCSS removes `//` but preserves `/**...*/`-style comments in output CSS, probably because only `/**...*/` ones are valid CSS.

In most cases these comments are irrelevant in CSS because they document SCSS variables, functions and mixins which are no longer there.

And SCSS compiler provides no option to drop comments: https://github.com/sass/node-sass/issues/1567.

I think we'd better stay with `//`

madoar replied:
I agree

madoar commented on src/css/archwizard.scss:
```diff
@@ -1,9 +1,97 @@
+//
+// Notes:
+// - The order of wizard step states in this mapping does not affect the order of resulting CSS rules, which is fixed
+//   internally by angular-archwizard.
```
Maybe also add a link or a short description where the order is fixed internally by angular-archwizard

madoar commented on src/css/archwizard.scss:
```diff
@@ -1,9 +1,97 @@
+//   internally by angular-archwizard.
+// - Depending on the active indicator style, some colors may not be used.  For example, 'background-color-default' and
+//   'background-color-hover' do not make sense in 'large-empty-symbols' indicator style.
+// - You can set a color to `null` to exclude the corresponding property from CSS output.
```
Where should I set the color to `null` to exclude the property from the CSS output?

earshinov replied:
Added an example in f57cf94

madoar commented on src/css/archwizard.scss:
```diff
@@ -1,9 +1,97 @@
+$wz-colors: (
+  '_': (
+    'default': (
+      'border-color-default': #E6E6E6,
```
I think we need some place where we document what each of these values means/defines. E.g.:

```
'border-color-default': The border color used for a wizard step in the default state (i.e. unselected and not yet completed)
...
```

earshinov replied:
I think it will result in too much comments and make it hard to see the code among them. The `$wz-colors` object is already elaborately documented, so from here

```
$wz-colors: (
  '_': (
    'default': (
      'border-color-default': #E6E6E6,
```

it should be understandable that '_' stands for any indicator style and 'default' is a wizard step state. What 'border-color' means is obvious. I added a paragraph about '-hover' and '-default' colors in 802e20f.

madoar replied:
Ok

madoar commented on src/css/archwizard.scss:
```diff
@@ -1,9 +1,97 @@
+$wz-label-color-hover: darken($wz-label-color-default, 20%) !default;
 
 // dot definitions
 $wz-dot-border-width: 2px !default;
```
Again we need some place where we define what a `dot` is and what the `dot width` is

madoar commented on src/css/wizard-navigation-bar.scss:
```diff
@@ -13,8 +13,8 @@
+$wz-param-step-state: null !global;
 
 
 @mixin wz-horizontal-line($dot-width, $dot-height, $line-color) {
```
Maybe we should add some comments to the mixins which contain what the mixins do

madoar replied:
I think you missed this comment

earshinov replied:
Yes, missed it, thanks! Comments added in 32b3c1e

madoar commented on src/css/wizard-navigation-bar.scss:
```diff
@@ -140,120 +120,155 @@ $wz-param-indicator-color: null !global;
       }
 
       $wz-param-indicator-state: 'default' !global;
```
I may have asked this before, but what exactly does `!global` do?

madoar replied:
I think you missed this comment

earshinov replied:
Yes, missed this comment, my bad :( `!global` tells SCSS that we are referencing the global variable `$wz-param-indicator-state`. Without this flag, SCSS would introduce a new local variable with the same name, which would not be visible in the mixin included below.

madoar commented on src/css/wizard-navigation-bar.scss:
```diff
@@ -140,120 +120,155 @@ $wz-param-indicator-color: null !global;
-    @include wz-define-style($layout) {
-      @include wz-state-circle-with-border-and-content($wz-param-indicator-color);
+// Define wizard styles specific to the configurated layouts (`$wz-layouts`) and indicator styles (`$wz-indicator-styles`)
+@mixin wz-define-styles() {
```
This mixin is kind of like black magic to me. I'm not that experienced with css and even less with scss, can you add some more content to the description/comment of the mixin to make it more clear what it does

earshinov replied:
Added comments in d2274c9

madoar commented on src/css/wizard-navigation-bar.scss:
```diff
@@ -140,120 +120,155 @@ $wz-param-indicator-color: null !global;
+//
+// Arguments:
+//   $indicator-style: Indicator style: 'small', 'large-filled' etc.
+//   $step-state:      Wizard step state: small', 'large-filled' etc.
```
Are you sure `$step-state` takes the same values as `$indicator-style`?

earshinov replied:
Definitely not :-) Fixed the comment in 5a9fd8a

madoar commented on src/css/wizard-navigation-bar.scss:
```diff
@@ -140,120 +120,155 @@ $wz-param-indicator-color: null !global;
+//     - the corresponding color definitions map entry is explicit `null`
+//
+@function wz-get-color($indicator-style, $step-state, $indicator-state, $property) {
+  $maps: (
```
This is the first time I've seen a map in `scss`. Can you elaborate a bit more about what `$maps` contains after defining it? And maybe also what exactly you access in the `$each` loop?

earshinov replied:
Yes, I added comments in 328fa26

madoar commented on src/css/wizard-navigation-bar.scss:
```diff
@@ -291,119 +306,89 @@ aw-wizard-navigation-bar {
+      width: 100%;
+      list-style: none;
+
+      @for $number-of-components from 2 through 10 {
```
If we are now doing this dynamically does it make sense to continue retaining the limit of `10`?

earshinov replied:
I think yes. If people actually needed more than 10 steps, there would be feature requests for it.

madoar commented on src/css/wizard-navigation-bar.scss:
```diff
@@ -291,119 +306,89 @@ aw-wizard-navigation-bar {
+      width: 100%;
+      list-style: none;
+
+      @for $number-of-components from 2 through 10 {
```
Does it make sense to use `1` as the minimum count? I don't see a reason why there shouldn't be a usecase where a wizard contains only a single step...

earshinov replied:
Wizard with a single step already looks perfectly without additional styles enabled by this `@for`. I have just created a demo page in angular-archwizard-demo project: https://github.com/madoar/angular-archwizard-demo/pull/49/commits/e205af439dfcb0771af9f4b33a354fb8aa14c409

madoar replied:
:+1:

earshinov commented:
> Please bear with some of my more silly questions. I'm not that knowledgeable with SCSS, therefore I've added some questions about parts of the code that I don't (fully) understand.

That's good. I think it is essential to make the code understandable to the project maintainer at the very least :-) I have added a lot of comments. Please review.

madoar commented on src/css/archwizard.scss:
```diff
@@ -1,11 +1,127 @@
+
+// Color definitions
+//
+// This is a mapping from indicator style and wizard step state to colors.
```
`This` as in `the color definitions`?

earshinov replied:
In 6202716 I squashed the two lines into

```
// Color definitions - a mapping from indicator style and wizard step state to colors.
```

This way it is more concise.

madoar commented on src/css/archwizard.scss:
```diff
@@ -1,11 +1,127 @@
 // dot definitions
+//
+// Here and below, "dots" are wizard step indicators represented as circles in the navigation bar.
+// Depending on step indicator style selected for the wizard, dots can either be small (indicator style 'small')
```
Depending on **the** step indicator style...

earshinov replied:
Fixed in ab5bfe9

earshinov commented:
> Your introduced variables start with $wz-... I guess wz stands for wizard right? Wouldn't it be better and more in line with our other definitions if the variables start with $aw (i.e. archwizard)?

I just followed the convention I saw in *.less files where variables had $wz-... prefix as well.

I agree that '$aw' is more appropriate. Let's rename.

Also, I would rename 'dots' to 'circles'.

madoar commented on README.md:
```diff
@@ -630,21 +630,43 @@ invalid state, which may lead to strange and unexpected behavior.
 ## Styles Customization
 
-If you are using SCSS, you can easily customize wizard's global styles and color theme using SCSS variables:
+If you are using SCSS, you can customize wizard's global styles and color theme using SCSS variables:
```
you can customize **the** wizard's global styles and color theme...

earshinov replied:
Fixed in 65ba9406a3c02f8bd78029f78fef0a2290f02253

madoar commented on src/css/wizard-navigation-bar.scss:
```diff
@@ -133,127 +117,182 @@ $wz-param-indicator-color: null !global;
-    }
-  }
+// Define wizard styles specific to the configurated layouts (`$aw-layouts`) and indicator styles (`$aw-indicator-styles`)
+@mixin aw-define-styles() {
```
Maybe add an empty line above the mixin comment to separate it better visually.

earshinov replied:
But there are already two blank lines there. As far as I can see, three blank lines are never used anywhere in archwizard
s SCSS files. In my opinion, using three blank lines here would be inconsistent.

earshinov commented:
> I think I prefer to add this PR to a new major release (i.e. 6.0.0) because it can lead to breaking changes for users with custom styles. @earshinov what do you suggest?

Yes, I think this change is worth a new major version.
